rfc_samplers.md

```diff
@@ -0,0 +1,125 @@
+
+Feature space
+-------------
+
+At a high level, a sampler returns a sequence of "clips" where a clip is a batch
+of (non-necessarily contiguous) frames. Based on an analysis of the sampling
+options of other libraries (torchvision, torchmultimodal, and various internal
+libraries), we want to support the following use-cases:
+
+
+**Clip location** (and number thereof)
+
+- Get `num_clips_per_video` random clips: typically called "Random" sampler
+- Get `num_clips_per_video` equally-spaced clips: "Uniform" sampler. Space is
+  index-based.
+- Get `num_clips_per_second` clips per seconds [within the [`start_seconds`,
+  `stop_seconds`] sampling range]: "Periodic" sampler.
+
+**Clip content**
+
+- Clip must be `num_frames_per_clip` frames spaced by `dilation` frames.
+
+Notes:
+
+- There are 3 sampling strategies defining the clip locations: Random, Uniform
+  and Periodic. The names are prone to confusion. "Random" is in fact uniformly
+  random while "Uniform" isn't random. "Uniform" is periodic and "Periodic" is
+  uniformly spaced too.
+- The parametrization of the *Clip Content* dimension is limited and
+  reduces to just `num_frames_per_clip` and the index-based `dilation`.
+- The *Clip Content* dimension is never random. `num_frames_per_clip` is always
+  passed by the user. This makes sense: training typically requires fixed-sized
+  inputs so `num_frames_per_clip` is expected to be constant. The content of a
+  clip (space between frames) is always regular, never random.
+
+
+Proposal
+--------
+
+```py
+from torchcodec.decoders import SimpleVideoDecoder
+from torchcodec import samplers
+
+decoder = SimpleVideoDecoder("./cool_vid")
+
+
+def samplers.clips_at_random_timestamps(
+    decoder,
+    *,
+    num_clips,
+    num_frames_per_clip=1,
+    seconds_between_frames=None,
```

Comment:
The output of this function is a sequence of clips where a clip is a batch of **frames**. A frame is a well-defined concept within the scope of these samplers, so I don't think we need to talk about timestamps here.

Reply:
Thinking about this more (refer to chat we had):

For time-based samplers the user has to give us the clip *duration* AND frames per clip both.

Your implementation will first determine the clip_starts in terms of seconds. To get the clip_ends you simply add the clip_duration (passed by the user) to the clip_starts.

Then to get the frames, you can subdivide the [clip_start, clip_end) interval into K segments and ask for the frames at the beginning of each segment using `get_frame_displayed_at`.

I think that's how this and the `at_regular_timestamps` below should work.

WDYT?

Reply:
I think users don't think in terms of clip duration. To them, a "clip" is something they pass to their model for training, which means that all clips must have a constant number of frames. So the _time duration_ of a clip is not important to them, but the _number of frames_ is.

What I think users think about when sampling from a video:

* How many frames are in each clip?
* How do I determine where each clips _starts_?
* Inside of each clip, how far apart should each frame be?

In this RFC, that last question is answered by `seconds_between_frames` and `num_indices_between_frames`. Specifying the distance between frames (in indices or time) means we don't need to specify the clip duration.

In terms of usability, I think specifying how far apart individual frames should be scales easier than clip duration. That is, when video length can vary by order of magnitudes, I think it's harder to specify a clip duration that makes sense across that spectrum.

Reply:
I agree with @scotts 

However to make the naming consistent I think it should be called seconds_between_samples or something like that. A frame is an *interval* not a *snapshot in time*. What the user wants is seconds between samples chosen and number of samples.

Note that if you specify two of {seconds_between_samples, num_samples_in_clip, clip_duration} the third is constrained to a single value. So specifying any two is enough. I suggested the user give us clip_duration and num_samples_in_clip, but the more intuitive may be num_samples_in_cip and seconds_between_samples

The last step is to get the frames that are displayed at those samples.

(I used samples instead of frames because two samples could land in the same frame, in which case we will be returning two copies of the same frame)

Reply:
To give an extreme example -- if the whole video is a single frame of 10 seconds, and if the user specifies seconds_between_frames=1, you would end up returning copies of the same frame in the clip. And there would not be 1 second between frames.

There would be 1 second between the *samples* (which are instants in time) the sampler chose -- and those samples just happen to land on the same frame.

Reply:
Let's discuss more today about the right parametrization (clip_duration vs seconds_between_frames).

Regarding using the term frame vs sample vs timestamp: let's try to leave that for a separate, higher-level discussion.